### ISSUE TITLE
array-callback-return incorrectly assumes an array

### ISSUE THREAD

**Issue by xPaw** [eslint/eslint#8428]
**Tell us about your environment**

* **ESLint Version:** v4.0.0-alpha.0
* **Node Version:** v7.6.0
* **npm Version:** v4.1.2

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:** https://github.com/thelounge/lounge/blob/8c987e7a2812dcc9f01df639df3d7c68339ab742/.eslintrc.yml

Reduced test case:
```js
require("event-stream").map(() => {});

var es = require("event-stream");
es.map(() => {});
```

This does not report an error on v3.19.0, probably due to `array-callback-return` not being in the `eslint:recommended`.

**Comment by Aladdin-ADD** [comment]
## <a name="eslint-recommended-changes"/> `eslint:recommended` changes

Three new rules have been added to the [`eslint:recommended`](http://eslint.org/docs/user-guide/configuring#using-eslintrecommended) config:

* [`array-callback-return`](/docs/rules/array-callback-return) requires predicate functions to return a value when using functional patterns such as `map` and `filter`.
* [`no-compare-neg-zero`](/docs/rules/no-compare-neg-zero) disallows comparisons to `-0`
* [`no-useless-escape`](/docs/rules/no-useless-escape) disallows uselessly-escaped characters in strings and regular expressions

**To address:** To mimic the `eslint:recommended` behavior from 3.x, you can disable these rules in a config file:

```json
{
  "extends": "eslint:recommended",

  "rules": {
    "array-callback-return": "off",
    "no-compare-neg-zero": "off",
    "no-useless-escape": "off"
  }
}
```

**Comment by xPaw** [comment]
@Aladdin-ADD I understand that. The problem is that rule incorrectly assumes that a function called `map` comes from `Array.prototype`.

**Comment by Aladdin-ADD** [comment]
@xPaw I think it is expected. As ESLint is a static code linting tool, it doesn't know the varialbe's real type.

**Comment by platinumazure** [comment]
Thanks for the issue.

Yes, `array-callback-return` is limited due to ESLint's nature as a static analysis tool and these sorts of false positives are not unexpected.

You have two options in these cases:
1. Use `// eslint-disable-line` or similar comments for cases where the rule is just wrong
1. Turn off the rule if it's just not working for you

That said, if you can identify a way to make the rule smarter, you are welcome to submit a pull request! And if you think an enhancement to support new user configuration might help, feel free to write a rule change proposal for that.

**Comment by xPaw** [comment]
Perhaps it could detect if the variable is assigned from a module require and ignore the error for that?

**Comment by not-an-aardvark** [comment]
> Perhaps it could detect if the variable is assigned from a module require and ignore the error for that?

I don't think that would be a good way to filter the errors, because in some cases a module require could actually return an array.

This was discussed in https://github.com/eslint/eslint/issues/4953#issuecomment-171990264 when `array-callback-return` was being considered for `eslint:recommended` for 2.0.0. At the time, it wasn't added due to concerns like this.

Out of interest, why are you calling `.map` with a noop function? I'm not familiar with `event-stream`.

**Comment by xPaw** [comment]
That noop was just a reduced testcase that triggers the error.

**Comment by not-an-aardvark** [comment]
Right, but is there actually a use case for not returning a value from a function like that?

**Comment by xPaw** [comment]
Event stream map is async and relies on a callback being called.

**Comment by not-an-aardvark** [comment]
Oh I see, [here](https://www.npmjs.com/package/event-stream#map-asyncfunction)'s an example from the docs. Thanks for explaining.

So it has a method called `map`, but it doesn't expect a value to be returned from the map function (it passes a callback instead).

`event-stream` is a very popular library, so I think we should probably remove `array-callback-return` from `eslint:recommended`. We removed `no-template-curly-in-string` in https://github.com/eslint/eslint/issues/8236 for much less.

**Comment by kaicataldo** [comment]
@not-an-aardvark Agreed - I think we should remove it too. `eslint:recommended` should be as generic as possible.

**Comment by platinumazure** [comment]
:+1: to removing from eslint:recommended from me.

**Comment by platinumazure** [comment]
TSC Summary: With the release of 4.0.0-alpha.0, users are confused about the `array-callback-return` rule when it flags false positives. A question has been raised about whether a rule that can cause confusion via false positives should be in eslint:recommended.

TSC Question: Should we remove array-callback-return from eslint:recommended?

----

If TSC comes to consensus before the meeting, feel free to remove from the agenda.

**Comment by btmills** [comment]
The TSC decided to remove `array-callback-return` from `eslint:recommended` in the 2017-04-13 meeting.
